<a href="https://colab.research.google.com/github/aguinaldoabbj/minicourse_flisol_natal_2019/blob/master/1_intro_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas basics




## Introduction

Pandas is a library that unifies the most common workflows that data analysts and data scientists previously relied on many different libraries for. Pandas has quickly became an important tool in a data professional's toolbelt and is the most popular library for working with tabular data in Python. Tabular data is any data that can be represented as rows and columns. 

To represent tabular data, Pandas uses a custom data structure called a **DataFrame**. A DataFrame is a highly efficient, 2-dimensional data structure that provides a suite of methods and attributes to quickly explore, analyze, and visualize data. The DataFrame is similar to the NumPy 2D array but adds support for many features that help you work with tabular data.

One of the biggest advantages that Pandas has over NumPy is the **ability to store mixed data types** in rows and columns. Many tabular datasets contain a range of data types and Pandas DataFrames handle mixed data types effortlessly while NumPy doesn't. Pandas DataFrames **can also handle missing values gracefully** using a custom object, **NaN**, to represent those values. A common complaint with NumPy is its lack of an object to represent missing values and people end up having to find and replace these values manually. In addition, pandas DataFrames contain axis labels for both rows and columns and enable you to refer to elements in the dataframe more intuitively. Since many tabular datasets contain column titles, this means that dataframes preserve the metadata from the file around the data.

To get things started, we have to load Pandas library:

In [0]:
import pandas as pd

## Our case study

In this part of the course, you'll learn the basics of pandas while exploring several datasets [Dados Abertos Natal](http://dados.natal.br/). Let's start things out by using the "Bilhetagem" dataset , which refers to the bus service billing system in Natal. 

You can dowload it through this URL: [Bilhetagem Analítica 2018](http://dados.natal.br/dataset/4fad551d-4d3b-4597-b8d3-7e887e22332e/resource/ec5b95a3-7b93-4346-98f6-1bd013faa651/download/dados-be-2018-analitico.csv)

## Downloading Data

We can download data directly from  [Dados Abertos Natal](http://dados.natal.br/) to our workspace in Google Colab. We can use **wget** download utility (available in most systems) to get the job done (the exclamation mark in the beginning sinalyzes Colab to use system's utilities instead of Python interpreter)

In [5]:
!wget -c https://raw.githubusercontent.com/aguinaldoabbj/minicourse_flisol_natal_2019/master/data/dados-be-2018-analitico.csv
#wget -c http://dados.natal.br/dataset/4fad551d-4d3b-4597-b8d3-7e887e22332e/resource/0d4a0f0f-b8a5-46a0-acfe-bb9138c8eb9e/download/dados-be-2019-analitico.csv

--2019-04-27 01:26:17--  https://raw.githubusercontent.com/aguinaldoabbj/minicourse_flisol_natal_2019/master/data/dados-be-2018-analitico.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 416 Range Not Satisfiable

    The file is already fully retrieved; nothing to do.



We can check if file was downloaded using 'ls':


In [6]:
!ls -lah  dados-be*

-rw-r--r-- 1 root root 72K Apr 27 01:26 dados-be-2018-analitico.csv


# First steps with pandas

## Loading data from a CSV file
As we can see, downloaded data is a CSV file. CSVs are commonly used to store tabular data. In simple words, it contains table rows whose cells are separated by some token, like a comma or semicolon. Pandas provides the *read_csv()* function, which takes the path of the CSV file and produces a DataFrame representation of its data.

In [0]:
# This CSV file uses ';' as separator and is encodded in iso-8859-1, so make this explicit when loading
df = pd.read_csv("dados-be-2018-analitico.csv", encoding='iso-8859-1', sep=';', thousands=r'.') 

We can check if data was loaded successfully by calling Dataframe's head( ) method and see the first 5 rows:

In [8]:
df.head()

Mês    Empresa  Linha  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
0    1  CONCEIÇÃO     21             13493          9506            35944   
1    1  CONCEIÇÃO     22              4644          3371            13088   
2    1  CONCEIÇÃO     30              8519          5614            20580   
3    1  CONCEIÇÃO     31              9631          6020            23460   
4    1  CONCEIÇÃO     41              6556          4038            20388   

   Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
0              8671                        0            17138         7397   
1              3230                        0             8109         3157   
2              5346                        0             7265         3191   
3              7587                        0             7065         2922   
4              7924                        0             5704         2791   

   Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  
0            6542            43727           3965         3684  
1            2529            15363              0         1737  
2            4415            21507           1537         2500  
3            5491            22212            140         2478  
4            3883            13870              0         2296

### Exploring the dataframe

Now that we've read the dataset into a DataFrame, we can start using more Pandas DataFrame methods to explore the data. In the same way we have just used the head() to see the firts rows of the DataFrame, we can use the tail( ) method to see the last ones:

In [9]:
df.tail()

Mês            Empresa  Linha  Estudante_Cartao  Estudante_BT  \
944   10  SANTA MARIA - URB    561                61            27   
945   10      VIA SUL - URB     50             48362         10350   
946   10      VIA SUL - URB     51             30382          6422   
947   10      VIA SUL - URB     52             21965          5475   
948   10      VIA SUL - URB     65             14231          3355   

     Vale_Transporte  Integracao_Plena  Integracao_Complementar  \
944             2029               536                        0   
945            71459             22085                      215   
946            25362              9241                        0   
947            23329              7763                        0   
948            14359              4247                        0   

     Gratuito_Cartao  Gratuito_BT  Inteira_Cartao  Inteira_Especie  \
944              131            7             219              695   
945            18052         9226           14547            58611   
946            10020         7065            6933            29185   
947            10908         7764            6190            27879   
948             4541         2308            3658            13157   

     Tarifa_Social  Qtd_Viagens  
944            120          340  
945           8484         5172  
946           2293         3265  
947           2089         3120  
948              0         1739

**Tip:** It's always good to use *head()* or *tail()* functions to limit dataset outputs.

A new Dataframe containing 10 random samples of the original Dataframe can be created by using the sample( ) method:

In [10]:
df.sample(10)

Mês            Empresa  Linha  Estudante_Cartao  Estudante_BT  \
63     1    GUANABARA - URB   7501                92            66   
115    2          CONCEIÇÃO   3101              1059           610   
943   10  SANTA MARIA - URB    411              7779          1142   
77     1           REUNIDAS   9018                13            23   
390    5          CONCEIÇÃO    411              8595          2733   
78     1  SANTA MARIA - URB     24             12272          7761   
809    9    GUANABARA - URB    593              6239           602   
19     1        DUNAS - URB     33             27684         22323   
154    2    GUANABARA - URB   2369               728           434   
267    3  SANTA MARIA - URB     36             10641          6001   

     Vale_Transporte  Integracao_Plena  Integracao_Complementar  \
63               945                42                        0   
115             1988               841                        0   
943             7816              3693                        0   
77               274               447                        0   
390            11278              4861                        0   
78             32366             11712                        0   
809             2497               780                        0   
19             74603             20083                        0   
154             1900               671                        0   
267            15815              5896                        0   

     Gratuito_Cartao  Gratuito_BT  Inteira_Cartao  Inteira_Especie  \
63               189           22             342              289   
115              576          305             489             1729   
943             3605         1233            2131             7783   
77               377            9              27              215   
390             4934         1647            3127            11640   
78             10384         5396            8682            29716   
809             1160          433             622             3195   
19             20022         7201           15832            70311   
154              942          410             369             1608   
267             8545         3976            3774            18038   

     Tarifa_Social  Qtd_Viagens  
63               0           38  
115              0          313  
943           1052          741  
77              30           94  
390            502         1080  
78             960         3369  
809              0          969  
19            6722         5137  
154              0          341  
267            262         2180

By these Dataframe printouts, it is possible to overview data structure (rows, columns and cells). A better understanding of the study Dataframe can be obtained with a series of Pandas Dataframe methods:

In [11]:
#get the shape of the df
df.shape


(949, 14)

So we have 949 entries (rows) of 14 variables (columns).

In [15]:
# show the basic information about the dataset (columns, sizes, column types, memory usage, etc)
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 949 entries, 0 to 948
Data columns (total 14 columns):
Mês                        949 non-null int64
Empresa                    949 non-null object
Linha                      949 non-null int64
Estudante_Cartao           949 non-null int64
Estudante_BT               949 non-null int64
Vale_Transporte            949 non-null int64
Integracao_Plena           949 non-null int64
Integracao_Complementar    949 non-null int64
Gratuito_Cartao            949 non-null int64
Gratuito_BT                949 non-null int64
Inteira_Cartao             949 non-null int64
Inteira_Especie            949 non-null int64
Tarifa_Social              949 non-null int64
Qtd_Viagens                949 non-null int64
dtypes: int64(13), object(1)
memory usage: 165.7 KB


Now we know lots important dataframe feaures. Let's run some statistics:

In [37]:
# describe statistical information about the numerical columns
df.describe()

Mês        Linha  Estudante_Cartao  Estudante_BT  \
count  949.000000   949.000000        949.000000    949.000000   
mean     5.492097  1109.404636      13434.656481   5478.087460   
std      2.884410  2239.082157      12384.928078   5566.013515   
min      1.000000     2.000000          0.000000      0.000000   
25%      3.000000    39.000000       4205.000000   1388.000000   
50%      5.000000    75.000000       9649.000000   3806.000000   
75%      8.000000   599.000000      19631.000000   7761.000000   
max     10.000000  9018.000000      64442.000000  38952.000000   

       Vale_Transporte  Integracao_Plena  Integracao_Complementar  \
count       949.000000        949.000000               949.000000   
mean      21266.004215       6149.785037                74.061117   
std       19292.437061       6169.000595               167.319789   
min           0.000000          0.000000                 0.000000   
25%        6942.000000       1489.000000                 0.000000   
50%       15724.000000       4247.000000                 0.000000   
75%       31829.000000       9111.000000                66.000000   
max      114307.000000      35933.000000              1353.000000   

       Gratuito_Cartao   Gratuito_BT  Inteira_Cartao  Inteira_Especie  \
count       949.000000    949.000000      949.000000       949.000000   
mean       7298.124341   3421.245522     4537.363541     20951.744995   
std        6327.152596   2965.329864     3772.870057     17782.050821   
min           0.000000      0.000000        0.000000         0.000000   
25%        2923.000000    979.000000     1505.000000      7228.000000   
50%        5993.000000   2929.000000     3591.000000     17184.000000   
75%       10321.000000   4965.000000     6897.000000     30846.000000   
max       37702.000000  17586.000000    18331.000000     88285.000000   

       Tarifa_Social  Qtd_Viagens  
count     949.000000   949.000000  
mean      937.771338  1913.467861  
std      1709.226991  1283.980331  
min         0.000000     1.000000  
25%         0.000000   761.000000  
50%       152.000000  1788.000000  
75%      1193.000000  2826.000000  
max     17298.000000  5407.000000

# Basic concepts and operations in pandas

## Indexing

When you read in a file into a DataFrame, pandas uses the values in the first row (also known as the header) for the column labels and the row number for the row labels. Collectively, the labels are referred to as the index. DataFrames contain both a row index and a column index. Here's a diagram that displays some of the column and row labels for data:

![Pandas Datafrane](https://raw.githubusercontent.com/ivanovitchm/cba2018/master/1-intro-pandas/indexing.png)

The labels allow us to refer to values in the DataFrame, which we'll learn more about in the rest of this notebook.

## Data types

When you displayed individual rows, represented as Series objects, you may have noticed the text "dtype: object" after the last value. "dtype: object" refers to the data type, or dtype, of that Series. The object dtype is equivalent to the string type in Python. Pandas borrows from the NumPy type system and contains the following dtypes:

* "object" - for representing string values.
* "int" - for representing integer values.
* "float" - for representing float values.
* "datatime" - for representing time values.
* "bool" - for representing Boolean values.

When reading a file into a DataFrame, pandas analyzes the values and infers each column's types. To access the types - for each column, use the DataFrame.dtypes attribute to return a Series containing each column name and its corresponding type. It is also possible to specify column's types at the moment of reading data into Pandas by using the dtype attribute in read_csv(). Read more about data types on the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/).

Sometimes we can encounter the value - "NaN" - for representing Not a Number, in other words is value that pandas dosen't know how to represent, normally it's associated with missing values.

In [17]:
df.dtypes

Mês                         int64
Empresa                    object
Linha                       int64
Estudante_Cartao            int64
Estudante_BT                int64
Vale_Transporte             int64
Integracao_Plena            int64
Integracao_Complementar     int64
Gratuito_Cartao             int64
Gratuito_BT                 int64
Inteira_Cartao              int64
Inteira_Especie             int64
Tarifa_Social               int64
Qtd_Viagens                 int64
dtype: object

## Series
The Series object is a core data structure that Pandas uses to represent rows and columns. A Series is a labelled collection of values similar to the NumPy vector. The main advantage of Series objects is the ability to utilize non-integer labels. NumPy arrays can only utilize integer labels for indexing.
Pandas utilizes this feature to provide more context when returning a row or a column from a DataFrame. For example, when you select a row from a DataFrame, instead of just returning the values in that row as a list, pandas returns a Series object that contains the column labels as well as the corresponding values:
The Series object representing the first row looks like:

## Selecting a row
While we use bracket notation to access elements in a NumPy array or a standard list, we need to use the Pandas method loc[] to select rows in a DataFrame. The loc[] method allows you to select rows by row labels. Recall that when you read a file into a dataframe, pandas uses the row number (or position) as each row's label. Pandas uses zero-indexing, so the first row is at index 0, the second row at index 1, and so on.
If you're interested in accessing a single row, pass in the row label to the loc[] method. Also, Python will return an error if you don't pass in a valid row label. For example, the following line takes the seventh row:

In [21]:
# Series object representing the seventh row.
df.loc[7]

Mês                                1
Empresa                    CONCEIÇÃO
Linha                             71
Estudante_Cartao                7188
Estudante_BT                    5395
Vale_Transporte                19626
Integracao_Plena                4209
Integracao_Complementar            0
Gratuito_Cartao                10865
Gratuito_BT                     4075
Inteira_Cartao                  3608
Inteira_Especie                23923
Tarifa_Social                   3773
Qtd_Viagens                     2389
Name: 7, dtype: object

## Selecting Multiple Rows

If you're interested in accessing multiple rows of the DataFrame, you can pass in either a slice of row labels or a list of row labels, and pandas will return a DataFrame. Note that unlike slicing lists in Python, a slice of a DataFrame using .loc[] will include both the start and the end row:

In [22]:
# DataFrame containing the rows at index 3, 4, 5, and 6 returned.
df.loc[3:6]

Mês    Empresa  Linha  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
3    1  CONCEIÇÃO     31              9631          6020            23460   
4    1  CONCEIÇÃO     41              6556          4038            20388   
5    1  CONCEIÇÃO     59              8184          6086            22623   
6    1  CONCEIÇÃO     63             21918         14054            44118   

   Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
3              7587                        0             7065         2922   
4              7924                        0             5704         2791   
5              7332                        0            13514         4373   
6             14630                        0            15034         5771   

   Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  
3            5491            22212            140         2478  
4            3883            13870              0         2296  
5            4622            29075           4573         2650  
6            7765            39855           3182         3768

In [23]:
# DataFrame containing the rows at index 2, 5, and 10 returned
df.loc[[2,5,10]]

Mês    Empresa  Linha  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
2     1  CONCEIÇÃO     30              8519          5614            20580   
5     1  CONCEIÇÃO     59              8184          6086            22623   
10    1  CONCEIÇÃO    411              5208          2869            10397   

    Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
2               5346                        0             7265         3191   
5               7332                        0            13514         4373   
10              4330                        0             4723         1654   

    Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  
2             4415            21507           1537         2500  
5             4622            29075           4573         2650  
10            2676            11043            580         1095

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Select the last 10 rows of our dataframe and assign to the variable last_rows. #tip: use .iloc[] method.


In [0]:
# put your code in this cell

## Selecting Individual Columns

When accessing a column in a DataFrame, Pandas returns a Series object containing the row label and each row's value for that column. To access a single column, use bracket notation and pass in the column name as a string:

In [0]:
df['Empresa'].head()

0    CONCEIÇÃO
1    CONCEIÇÃO
2    CONCEIÇÃO
3    CONCEIÇÃO
4    CONCEIÇÃO
Name: Empresa, dtype: object

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)


1.   Assign the "Empresa" column to the variable empresas
2.   Assign the last 5 rows of "Linha" column to the variable linhas. #tip: use the tail() function.

In [0]:
# put your code in this cell

## Selecting Multiple Columns By Name and Column Index

To select multiple columns, pass in a list of strings representing the column names and pandas will return a dataframe containing only the values in those columns. The following code returns a dataframe containing the "Empresa" and "Linha" columns, in that order:

In [0]:
df[['Empresa','Linha']].head()

Empresa  Linha
0  CONCEIÇÃO     21
1  CONCEIÇÃO     22
2  CONCEIÇÃO     30
3  CONCEIÇÃO     31
4  CONCEIÇÃO     41

When selecting multiple columns, the order of the columns in the returned DataFrame matches the order of the column names in the list of strings that you passed in. This allows you to easily explore specific columns that may not be positioned next to each other in the DataFrame. You can do the same operation if you know the column index using iloc[ ]:

In [39]:
df.iloc[:,[1,2]].head()

Empresa  Linha
0  CONCEIÇÃO     21
1  CONCEIÇÃO     22
2  CONCEIÇÃO     30
3  CONCEIÇÃO     31
4  CONCEIÇÃO     41

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Use **df.columns** to do the same operation as above. 

In [0]:
# put your code in this cell

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Select and display at least a random sample row of a DataFrame created with only "Empresa", "Linha" and "Mês" columns. Tip: use the sample() or any() functions.

In [0]:
# put your code in this cell

## Selecting Rows by Criteria
It is also possible to select rows based on a search criteria. Let's say we want only data regarding the company "CONCEIÇÃO":

In [0]:
df[df['Empresa'] == "CONCEIÇÃO"].head()

Mês    Empresa  Linha  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
0    1  CONCEIÇÃO     21             13493          9506            35944   
1    1  CONCEIÇÃO     22              4644          3371            13088   
2    1  CONCEIÇÃO     30              8519          5614            20580   
3    1  CONCEIÇÃO     31              9631          6020            23460   
4    1  CONCEIÇÃO     41              6556          4038            20388   

   Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
0              8671                        0            17138         7397   
1              3230                        0             8109         3157   
2              5346                        0             7265         3191   
3              7587                        0             7065         2922   
4              7924                        0             5704         2791   

   Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  
0            6542            43727           3965         3684  
1            2529            15363              0         1737  
2            4415            21507           1537         2500  
3            5491            22212            140         2478  
4            3883            13870              0         2296

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Put in the variable "selected_bus_lines" the dataframe resulting from the selection of bus lines number "21" and "22".

**Tip:** You can use function *isin()* function for the target column.

In [0]:
# put your code in this cell

## Column Uniqueness

When in the process of knowing the data, it's interesting to know the uniqueness of some attibutes. For this, Pandas implements the function unique() which can be use, for example, to know how many different bus companies appear in our data:

In [0]:
# uniqueness of column "Empresa"
empresas = df['Empresa'].unique()


In [45]:
# length of the array returned
len(empresas)

6

So, there are 6 different bus companies in our DataFrame.

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

How many different bus lines are there in our DataFrame?

In [0]:
# put your code in this cell

## Data manipulation with Pandas

### Overview
In the previous sections, we learned how to explore a pandas DataFrame. In this mission, we'll explore how to manipulate a DataFrame and make transformations to it. We'll continue to work with the same data set from the Power Plants information. We'll build a better dataset cleaning the data and removing not useful information. We also gonna learn how to group up informatio and manipulate data.

## Dropping Columns
Sometimes datasets have more information than needed or bad quality data that may be discarded. Let's suppose we don't have to look at half-price entries for students (columns with *Estudante_* as prefix) and, to simplify things outs, let's drop them out of the dataset with the *drop()* function:

**Tip:** a list of dataset columns can be obtained by calling *df.columns*

In [47]:
df.drop(['Estudante_Cartao','Estudante_BT'], axis=1).head() # note that axis=1 means along with "columns". It's a column-wise operation.

Mês    Empresa  Linha  Vale_Transporte  Integracao_Plena  \
0    1  CONCEIÇÃO     21            35944              8671   
1    1  CONCEIÇÃO     22            13088              3230   
2    1  CONCEIÇÃO     30            20580              5346   
3    1  CONCEIÇÃO     31            23460              7587   
4    1  CONCEIÇÃO     41            20388              7924   

   Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  Inteira_Cartao  \
0                        0            17138         7397            6542   
1                        0             8109         3157            2529   
2                        0             7265         3191            4415   
3                        0             7065         2922            5491   
4                        0             5704         2791            3883   

   Inteira_Especie  Tarifa_Social  Qtd_Viagens  
0            43727           3965         3684  
1            15363              0         1737  
2            21507           1537         2500  
3            22212            140         2478  
4            13870              0         2296

The same way we dropped some columns, we could think of keeping only certain columns in the dataset. Supposing we want to use only columns "Mês", "Empresa" and "Qtd_Viagens", we can use the *difference()* function to help us out on this:

In [48]:
df.drop(df.columns.difference(['Mês','Empresa','Qtd_Viagens']), axis=1).head()

Mês    Empresa  Qtd_Viagens
0    1  CONCEIÇÃO         3684
1    1  CONCEIÇÃO         1737
2    1  CONCEIÇÃO         2500
3    1  CONCEIÇÃO         2478
4    1  CONCEIÇÃO         2296

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Create a new dataframe containing no data regarding subsidized entrance tickets according to the following recipe:


1.  Check column names using *df.columns.tolist()* and assign the resulting list to the variable **col_names**
2.  Remove undesired column names from variable **col_names** using the *remove()*  function.
3.  Use variable **col_names** to map only the desired columns to a new dataframe.

**Tip:** There is no use for the drop() function in this exercise.



In [0]:
# put your code in this cell

## Adding Columns

If instead of dropping, we want to add a new column to the dataframe? What about adding a new column called 'Integracao_Total', which results from the sum of columns 'Integracao_Plena' and 'Integracao_Complementar' ?



In [50]:
df['Integracao_Total'] = df['Integracao_Plena'] + df['Integracao_Complementar']
df.sample(5)

Mês          Empresa  Linha  Estudante_Cartao  Estudante_BT  \
877   10      DUNAS - URB    331             20027          3780   
121    2      DUNAS - URB    331             11588          5935   
664    8        CONCEIÇÃO     30             23907          7563   
710    8  GUANABARA - URB    102              2309           802   
390    5        CONCEIÇÃO    411              8595          2733   

     Vale_Transporte  Integracao_Plena  Integracao_Complementar  \
877            22162              4791                       16   
121            18454              3843                        0   
664            24858              6866                        0   
710             4507               922                       14   
390            11278              4861                        0   

     Gratuito_Cartao  Gratuito_BT  Inteira_Cartao  Inteira_Especie  \
877             6562         2571            5993            18942   
121             5098         1954            5291            17771   
664             8361         3627            6239            24311   
710             2588          994             989             5606   
390             4934         1647            3127            11640   

     Tarifa_Social  Qtd_Viagens  Integracao_Total  
877           2227         1560              4807  
121            932         1395              3843  
664              0         2646              6866  
710              0          433               936  
390            502         1080              4861

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Create a new column of your own by using any arithmetic operation involving the columns of the dataframe.



In [0]:
# put your code in this cell

## Column-wise Operations

A common step in data analysis is to apply an operation or a function to a specific entire column. A primary operation would be a type change in a column. 

Column "Linha" refers to the name of the bus line. Hence, it is plausible to convert the type of this column to string (str). We can do it with the help of the *astype()* function.

In [0]:
df["Linha"] = df["Linha"].astype(str)

In [53]:
df.dtypes

Mês                         int64
Empresa                    object
Linha                      object
Estudante_Cartao            int64
Estudante_BT                int64
Vale_Transporte             int64
Integracao_Plena            int64
Integracao_Complementar     int64
Gratuito_Cartao             int64
Gratuito_BT                 int64
Inteira_Cartao              int64
Inteira_Especie             int64
Tarifa_Social               int64
Qtd_Viagens                 int64
Integracao_Total            int64
dtype: object

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

As you can see in previous steps, Pandas guessed the type *int64* for all numeric columns. This means numbers are signed 64-bit integers. Is this precision necessary for the numerical magnitude of this dataset? If not, change the type of these columns for *int32*. After doing that, check the resulting memory economy.

**Tip:** Use df.columns.difference() to exclude columns from conversion.



In [0]:
# put your code in this cell

### The Apply Method

Sometimes it is needed to apply a more complex operation to one or more columns of the data set. To accomplish this more easily, pandas offers the *apply* method, which is able to apply a custom column-wise operation.

Column 'Mês' in the dataframe refers to month numbers. It is also interesting to have a column with the corresponding month names, don't you think? We can do it with the *apply* method, a [lambda function](https://www.w3schools.com/python/python_lambda.asp) and a helper library called **calendar** (to map month numbers into month names). The resulting list of month names could be added as a new column for the dataframe.

In [0]:
#lets import calendar
from calendar import month_name #we want only month_name from calendar.

Calendar's month_name is just an array with months names ordered by number. To get the names, we can just access an element by its index number. For example, if we access the 3rd element in the array, we get 'March'. 

In [56]:
month_name[3]

'March'

Thus, we can use apply this mechanism to the whole column "Mês" and obtain as result an array of month names:

In [57]:
month_names = df.Mês.apply(lambda x: month_name[x]) #making the lambda function column wise with apply
month_names.sample(5)

738    August
336     April
368     April
729    August
483      June
Name: Mês, dtype: object

In [58]:
# adding new column called 'Month' with the month names list generated from column 'Mês'
df['Month'] = month_names
df.head()

Mês    Empresa Linha  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
0    1  CONCEIÇÃO    21             13493          9506            35944   
1    1  CONCEIÇÃO    22              4644          3371            13088   
2    1  CONCEIÇÃO    30              8519          5614            20580   
3    1  CONCEIÇÃO    31              9631          6020            23460   
4    1  CONCEIÇÃO    41              6556          4038            20388   

   Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
0              8671                        0            17138         7397   
1              3230                        0             8109         3157   
2              5346                        0             7265         3191   
3              7587                        0             7065         2922   
4              7924                        0             5704         2791   

   Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  \
0            6542            43727           3965         3684   
1            2529            15363              0         1737   
2            4415            21507           1537         2500   
3            5491            22212            140         2478   
4            3883            13870              0         2296   

   Integracao_Total    Month  
0              8671  January  
1              3230  January  
2              5346  January  
3              7587  January  
4              7924  January

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Create a new column called "Mdays" containing the number of days for each month of column 'Mês'

**Tip:** Use mdays[ ] from library **calendar.** 



In [0]:
# put your code in this cell

## Sorting a dataframe by a column

A sorted dataframe is always better to figure things out. To do so, we need choose a column and to sort the DataFrame by that column. Why don't we build a to 10 rank the whole dataframe in relation to the round trips ("Qtd_Viagens" column)? We can use the method  *sort_values()* to help us out:

In [60]:
df.sort_values("Qtd_Viagens", ascending=False).head() # if ascending=False, the rank order will be descending

Mês        Empresa Linha  Estudante_Cartao  Estudante_BT  \
756    8  VIA SUL - URB    50             50046         18556   
921   10       REUNIDAS    73             64442         17125   
258    3       REUNIDAS    73             56313         38952   
730    8       REUNIDAS    73             59512         28088   
634    7       REUNIDAS    73             36317         23770   

     Vale_Transporte  Integracao_Plena  Integracao_Complementar  \
756            75385             24216                      270   
921           112219             34462                      552   
258           114307             35933                      572   
730           112950             35284                      561   
634           110321             31765                      551   

     Gratuito_Cartao  Gratuito_BT  Inteira_Cartao  Inteira_Especie  \
756            19818         8773           16290            63519   
921            17811        10888           16827            79980   
258            17999         9226           16100            85992   
730            17999        10048           17879            86623   
634            17278         9543           16770            82161   

     Tarifa_Social  Qtd_Viagens  Integracao_Total    Month  
756              0         5407             24486   August  
921          17298         5402             35014  October  
258           3742         5259             36505    March  
730              0         5224             35845   August  
634              0         5205             32316     July

We can also use more columns to sort the dataframe by. Let's now rank dataframe rows by columns regarding half-priced entries for students (coluns with "Estudante_" prefix):

In [61]:
df.sort_values(by=['Estudante_Cartao','Estudante_BT'], ascending=False).head(10) # by statement is needed when using more than one column

Mês      Empresa Linha  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
921   10     REUNIDAS    73             64442         17125           112219   
827    9     REUNIDAS    73             64051         14496           106976   
681    8  DUNAS - URB    33             63380         22643            76702   
874   10  DUNAS - URB    33             62959         11712            73935   
779    9  DUNAS - URB    33             60017          9664            67806   
730    8     REUNIDAS    73             59512         28088           112950   
668    8    CONCEIÇÃO    63             59056         17759            51840   
399    5  DUNAS - URB    33             56739         22310            73401   
305    4  DUNAS - URB    33             56526         28396            72578   
258    3     REUNIDAS    73             56313         38952           114307   

     Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
921             34462                      552            17811        10888   
827             33082                      544            16997        10421   
681             20822                      205            20232         7795   
874             20262                      182            20276         8003   
779             18870                      161            18092         7474   
730             35284                      561            17999        10048   
668             26449                        0            19201         7964   
399             20825                        0            19751         6701   
305             20750                        0            19562         6566   
258             35933                      572            17999         9226   

     Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  \
921           16827            79980          17298         5402   
827           16265            88285           4926         5107   
681           18331            71492              0         4575   
874           16988            64794          11973         4481   
779           16496            67102           2964         4305   
730           17879            86623              0         5224   
668           10033            45732              0         4303   
399           16488            70927           5240         4495   
305           16377            68632           3156         4693   
258           16100            85992           3742         5259   

     Integracao_Total      Month  
921             35014    October  
827             33626  September  
681             21027     August  
874             20444    October  
779             19031  September  
730             35845     August  
668             26449     August  
399             20825        May  
305             20750      April  
258             36505      March

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Build a rank containing **only** the column "Linha" and a target column (chosen by you). You can sort rows in any order.

**Tip:** Use function *df.drop() *do get rid of undesired columns and then use *df.sort_values()*.

In [0]:
# put your code in this cell

## Grouping information by a column

Sometimes we need to group information, to do so we can use the function *groupby()* to group data and then apply to the result an aggregation function. In this example,  we gonna group all dataframe in relation to the columns "Month" and "Mês" (because they represent the same information). The result will be a montly aggregation of all data. 

In [64]:
df.groupby(['Month','Mês'], as_index=False).sum() #as_index=False means that the aggregation column will be not turned into an index

Month  Mês  Estudante_Cartao  Estudante_BT  Vale_Transporte  \
0      April    4           1331095        678262          2016760   
1     August    8           1584763        560499          2164823   
2   February    2            922115        560997          1793071   
3    January    1            729392        521541          2010298   
4       July    7           1084662        484700          2024453   
5       June    6           1218904        542724          1980585   
6      March    3           1368779        743969          2103611   
7        May    5           1391679        512755          2061006   
8    October   10           1571000        319749          2066691   
9  September    9           1547100        273509          1960140   

   Integracao_Plena  Integracao_Complementar  Gratuito_Cartao  Gratuito_BT  \
0            604373                     7247           696498       312804   
1            640724                     8088           755925       364295   
2            504393                     5204           606608       275400   
3            549134                     5591           671461       305725   
4            557249                     7284           684655       318444   
5            570681                     7436           666944       311043   
6            622058                     6953           730598       323111   
7            608455                     7760           706790       319696   
8            601418                     7657           720123       366005   
9            577661                     7064           686318       350239   

   Inteira_Cartao  Inteira_Especie  Tarifa_Social  Qtd_Viagens  \
0          421292          1934271          80615       182136   
1          488153          2135562              0       195284   
2          370324          1857316          67013       161318   
3          404810          1998704         168228       184030   
4          430359          1961328              0       182527   
5          418158          1884429              0       174948   
6          445180          2106287          59430       187775   
7          436692          2016118         136120       178362   
8          449238          1946104         295491       189004   
9          441752          2043087          83048       180497   

   Integracao_Total  
0            611620  
1            648812  
2            509597  
3            554725  
4            564533  
5            578117  
6            629011  
7            616215  
8            609075  
9            584725

### Exercise

![alt text](https://cdn.dribbble.com/users/2344801/screenshots/4774578/alphatestersanimation2.gif =150x120)

Make an monthly account per bus company using *groupby()*.

**Tip:** You will need to convert column "Mês" to string to make this exercise easier.



In [0]:
# put your code in this cell